source/Body.cpp

```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
-		frameOffset += static_cast<float>(Random::Real()) * cycle;
+		integratedFrame = static_cast<float>(Random::Real()) * cycle;
 	}
-	else if(startAtZero)
+	else if(anim.startAtZero)
 	{
-		startAtZero = false;
-		// Adjust frameOffset so that this step's frame is exactly 0 (no fade).
-		frameOffset -= frameRate * step;
+		anim.startAtZero = false;
+		// Adjust integrated frame to start at the preferred start frame.
+		integratedFrame = anim.startFrame;
 	}
 
+	// Clamp frameRate from previous calculations.
+	if(frameRate < Body::MIN_FRAME_RATE)
+		frameRate = Body::MIN_FRAME_RATE;
+
+	if(frameRate > anim.frameRate)
+		frameRate = anim.frameRate;
+
 	// Figure out what fraction of the way in between frames we are. Avoid any
 	// possible floating-point glitches that might result in a negative frame.
-	frame = max(0.f, frameRate * step + frameOffset);
-	// If repeating, wrap the frame index by the total cycle time.
-	if(repeat)
-		frame = fmod(frame, cycle);
-
-	if(!rewind)
+	int prevFrame = static_cast<int>(frame), nextFrame = -1;
+	// Integrate rampRate in order to determine frame.
+	integratedFrame += frameRate * (step - currentStep);
+	frame = max(0.f, integratedFrame);
+	// Flagged for if a state transition has been completed this step.
+	bool stateTransitionCompleted = false;
+	if(!stateTransitionRequested)
 	{
-		// If not repeating, frame should never go higher than the index of the
-		// final frame.
-		if(!repeat)
-			frame = min(frame, lastFrame);
-		else if(frame >= frames)
+		// Handle any frameRate changes in the animation.
+		if(anim.rampUpRate != 0.0)
+		{
+			if(frameRate >= Body::MIN_FRAME_RATE && frameRate <= anim.frameRate)
+				frameRate += anim.rampUpRate;
+		}
+		else
+			frameRate = anim.frameRate;
+		// For when it needs to be applied to transition.
+		delayed = 0.f;
+
+		// If repeating, wrap the frame index by the total cycle time.
+		if(anim.repeat)
+			frame = fmod(frame, cycle);
+
+		if(!anim.rewind)
+		{
+			// If not repeating, frame should never go higher than the index of the
+			// final frame.
+			if(!anim.repeat)
+			{
+				frame = min(frame, lastFrame);
+				framePercentage = (frame + 1) / frames;
+				if(framePercentage >= anim.indicatePercentage)
+					stateReady = anim.indicateReady;
+				else
+					stateReady = false;
+			}
+			else if(frame >= frames)
+			{
+				stateReady = false;
+				// If in the delay portion of the loop, set the frame to zero.
+				frame = 0.f;
+			}
+		}
+		else if(frame >= lastFrame)
 		{
-			// If we're in the delay portion of the loop, set the frame to 0.
-			frame = 0.f;
+			// In rewind mode, once you get to the last frame, count backwards.
+			// Regardless of whether we're repeating, if the frame count gets to
+			// be less than 0, clamp it to 0.
+			frame = max(0.f, lastFrame * 2.f - frame);
+			stateReady = false;
 		}
+		nextFrame = static_cast<int>(frame);
+		if(nextFrame != prevFrame)
+			randomFrame = static_cast<int>(static_cast<float>(Random::Real()) * frames);
```

Comment:
(see my previous comment)
```suggestion
		nextFrame = frame;
		if(nextFrame != prevFrame)
			randomFrame = Random::Real() * frames;
```